Add Bash and Ansible checks for versions

jan-cerny opened the pull request:
#### Description:
- add Ansible and Bash version checks to package CPE platform
- remove exception that prevents adding Bash and Ansible version checks
- add unit test for Bash macro used in the aforementioned check


#### Rationale:
Enable usage of versioned platform expressions in rules.

Fixes: #9983

#### Review Hints:
Add platform package expressions to your favorite rule, eg.: `platform: package[sudo]<=8.4`, then build the data stream and check the generated Bash and Ansible expressions in the built data streams.

jan-cerny commented:
@jan-cerny Could you please document what version formats are supported by these checks?

matejak commented on shared/macros/10-ansible.jinja:
```diff
@@ -700,12 +700,20 @@ The following macro remediates Audit syscall rule in :code:`/etc/audit/audit.rul
   {{%- set package = platform_package_overrides[package] -%}}
 {{%- endif -%}}
+{{%- if ver -%}}
+"{{{ package }}}" in ansible_facts.packages and (ansible_facts.packages["{{{ package }}}"] | last)["version"] is version("{{{ ver }}}", "{{{ op }}}")
```
What is that `last` thing here?

jan-cerny replied:
It chooses the last item form a list. The `packages` dictionary values are lists, because there can be multiple packages of the same name installed on a system.

matejak replied:
How interesting! I think that this is just a theoretical possibility, so handling it like this seems correct.

matejak commented:
Thank you for this effort, and also for tests. However, first and foremost, I miss the definition of the check's expected behavior (at least related to the OVAL code), and the human-readable description of the algorithm in the PR. Ideally, this should be understandable in the code form, but Jinja is not a very expressive language, so it may be necessary to describe it by means of comments. Ansible is quite simple, so it is mostly about Bash.

Next, I would like to have a way of testing the Bash algorithm without the RPM layer, i.e. using a mock-like approach, so one may perform the fine-grained close-to-the-edge tests easily.

jan-cerny commented:
@matejak  I have added documentation and description of the checks in Bash macros. I have changed the test to use a mock function instead of the real `rpm` command to avoid dependency on a specific rpm package being present on the test worker machine. I have rebased the PR on the top of the latest upstream branch. Do you have any suggestions how to achieve the definition of the check's expected behavior?

matejak commented on shared/macros/10-bash.jinja:
```diff
@@ -1875,40 +1875,133 @@ Part of the grub2_bootloader_argument_absent template.
+:param ver: package version
+:param op: version comparison operator ("<", "<=", "==", "!=", ">", ">=")
+#}}
+{{%- macro bash_compare_version_rpm(package, ver, op) -%}}
```
You couple two things here - retrieval of the version and the comparison algorithm.
Why not to redesign it to have
1. A macro `bash_compare_version` that accepts a Bash string,
2. Wrapping it in a macro `bash_compare_version_rpm` that would call the former with `$({{{ bash_get_rpm_package_version(package) }}})` as an argument?

That way, you won't have to mock the `rpm` command in tests.

matejak commented:
> @matejak I have added documentation and description of the checks in Bash macros. I have changed the test to use a mock function instead of the real `rpm` command to avoid dependency on a specific rpm package being present on the test worker machine. I have rebased the PR on the top of the latest upstream branch. Do you have any suggestions how to achieve the definition of the check's expected behavior?

Technically, the algorithm description and pointing people to the tests should be good enough, as tests are the best definition of specification, at least if they are exhaustive enough to cover all of the interesting cases.

matejak commented on tests/unit/bash/bash_pkg_conditional_rpm.bats.jinja:
```diff
@@ -0,0 +1,76 @@
+    {{{ bash_pkg_conditional_rpm("coconut", "5.0.0-1", "<=") | indent(4) }}}
+    {{{ bash_pkg_conditional_rpm("coconut", "5.0.0-1.fc31", "<=") | indent(4) }}}
+
+    {{{ bash_pkg_conditional_rpm("coconut", "4.5.17-9.fc31", "<") | indent(4) }}}
```
I don't like the `.fc31` bit, I think that a content author shouldn't have to worry about it at all.
I am even tempted to suggest to drop the release part - releases stop to be new very fast, then they become common, and even superseded. IOW, if one's system is up-to-date, then the release should be "high enough" at scan-time, as a release that exists at content-writing time will have time to propagate to individual systems.

matejak commented on tests/unit/bash/bash_pkg_conditional_rpm.bats.jinja:
```diff
@@ -0,0 +1,76 @@
+    {{{ bash_pkg_conditional_rpm("coconut", "4.1.1-9.fc31", ">=") | indent(4) }}}
+    {{{ bash_pkg_conditional_rpm("coconut", "4.1.1-1.fc31", ">=") | indent(4) }}}
+
+    {{{ bash_pkg_conditional_rpm("coconut", "4.5.17-8.fc31", "<=") | indent(4) }}}
```
I expect the usage `coconut>=4.5` to be the most common - if you have an upstream version that is known to have a feature, then it is the upper bound of all version numbers that may have it, as a feature may be backported to an eariler version.
Therefore, various distributions could specify their distribution-specific version specifiers if they are after it, and a catch-all spec would look like a safe comparison against the upstream version, not considering the release number and maybe even not the minor number.

Anyway, I would add such example here, the good thing is that it works.

matejak commented on tests/unit/bash/bash_pkg_conditional_rpm.bats.jinja:
```diff
@@ -0,0 +1,76 @@
+    {{{ bash_pkg_conditional_rpm("coconut", "3.0.0-1", ">") | indent(4) }}}
+    {{{ bash_pkg_conditional_rpm("coconut", "3.0.0-1.fc31", ">") | indent(4) }}}
+
+    {{{ bash_pkg_conditional_rpm("coconut", "3", ">") | indent(4) }}}
+    {{{ bash_pkg_conditional_rpm("coconut", "3.0", ">=") | indent(4) }}}
+    {{{ bash_pkg_conditional_rpm("coconut", "3.0.0-1", ">=") | indent(4) }}}
+    {{{ bash_pkg_conditional_rpm("coconut", "3.0.0-1.fc31", ">=") | indent(4) }}}
```
This seems to be a double copy-paste.

matejak commented:
The Bash code probably doesn't handle epochs well. Unfortunately, the `rpm` command outputs the epoch as `(none)` instead of e.g. 0, which somehow complicates the comparison. We have to think how to handle this case.

jan-cerny commented on linux_os/guide/services/kerberos/kerberos_disable_no_keytab/rule.yml:
```diff
@@ -29,8 +29,7 @@ references:
 
-platforms:
-    - krb5_server_older_than_1_17-18 and krb5_workstation_older_than_1_17-18
+platform: package[krb5-server]<1.17-18 and package[krb5-workstation]<1.17-18
```
@matejak Do you recommend to move the changes of real use of the feature (ie. changing platform to use versioned expressions) to a separate PR?

jan-cerny replied:
@matejak ^

vojtapolasek commented on linux_os/guide/services/kerberos/package_krb5-server_removed/rule.yml:
```diff
@@ -32,8 +32,7 @@ references:
 
-platforms:
-    - krb5_server_older_than_1_17-18
+platform: package[krb5-server]<1.17.18
```
I have noticed an interesting thing which might be a bug.
Note that here you use the version in the form 1.17.18. But in other cases you use 1.17-18. This is inconsistent and needs to be fixed... but that is not the problem this comment is about.
It seems that there is a problem with generating names of CPE names - in this state it should generate two independent CPE names (one for 1.17-18 and one for 1.17.18). But only one name is generated.
This should not get fixed in this PR probably, but it should get fixed eventually.

jan-cerny replied:
nice catch! thanks!

vojtapolasek commented:
Hello @jan-cerny , I have found another problem, this time in Ansible.
Ansible package facts have the version split into two values - the version (which you are using now) and release. -You will probably need to concatenate those strings with - to achieve the same behavior as with the Bash conditional.

matejak commented:
We have found out that the default Ansible comparison is based on the `LooseVersion` concept - see https://epydoc.sourceforge.net/stdlib/distutils.version.LooseVersion-class.html
As a result, it is clear that all algorithms will produce the same result when we talk about versions as dot-separated tuple of just numbers- let's call this as numeric version. We can't control packages on the scanned system that may have non-numeric versions, but we can raise a build-time error if the reference version specified by the content author is non-numeric with a notice that only numeric versions are supported.
Generally, as we target updated systems anyway, we don't have to rely on fine comparisons of package versions, and we can use system versions instead, e.g. something like rhel>8.6 instead of foo>1.2.3p1, as we know that an up-to-date rhel 8.6 and later comes with foo at least 1.2.3p1.
Which leaves us with these two things:
1. We need to document limitations of the version handling and act on those limitations if they are violated during build-time, and
2. We need to check out how to deal with epochs, as epochs are common, and they were introduced mostly because the standard versioning ceased to be continuous.

matejak commented on tests/unit/bash/bash_pkg_conditional_rpm.bats.jinja:
```diff
@@ -0,0 +1,77 @@
+}
+
+@test "bash_pkg_conditional_rpm - test package version" {
+    {{{ bash_pkg_conditional_rpm("coconut", "5", "<") | indent(4) }}}
```
Please consider introducing a macro that is rpm-agnostic, and that would be called as
`{{{ bash_pkg_conditional_compare("4.5.17-8", "5", "<") | indent(4) }}}`

jan-cerny commented:
@matejak Thanks for the great summary! I will look into what we can do.

jan-cerny commented:
> Let's drop the whole release thing, because before our content gets to scanned systems, packages there are supposed to be updated to the latest release available.

I think that it can be easily done by adjusting the code macros and dropping the part behind `-` if it exists.

> We need to document limitations of the version handling and act on those limitations if they are violated during build-time, and

I think that we can put a check to the template preprocess hook Python function.

> We need to check out how to deal with epochs, as epochs are common, and they were introduced mostly because the 
standard versioning ceased to be continuous.

This will be the most difficult item. At this moment, the boolean expression parser doesn't support using colons (`:`) in the expressions, but a colon is traditionally used as a separator of the epoch in the EVR versions strings. If you add a platform eg. `platform: package[sudo]>1:4.8` to your rule  you will get an exception `ssg.ext.boolean.boolean.ParseError: Unknown token for token: ":" at position: 15`  The boolean expression parser is an external library.

I  don't know how to address this one. Any suggestions?

jan-cerny commented:
I have changed the logic to not use the release part of the version strings. I have add a build-time exception to the platform_package template checking the format of the version string used in the platform expression. I have updated the documentation. I have made the `bash_pkg_conditional_rpm` macro independent from the`rpm` command. I have also rebased the PR on the top of the latest upstream master branch.

vojtapolasek commented:
@jan-cerny  it is easy to enable the build system to accept ":" in the version string. Just add the character to this list:
https://github.com/ComplianceAsCode/content/blob/de8dc7c0a01170fa56b2a40c77803144290c0f79/ssg/boolean_expression.py#L6

vojtapolasek commented:
@jan-cerny @matejak  I tested this approach when we leave out the release part of the version specifier and I found something that needs to be solved.
I tested only OVAL behavior so far, but I believe it is enough to consider it.
If the current implementation encounters a version specifier without the release part, it automatically fills in "0" as the release. The relevant function is here:
https://github.com/ComplianceAsCode/content/blob/de8dc7c0a01170fa56b2a40c77803144290c0f79/ssg/utils.py#L90
This could be, of course, simply fixed. But consider my following test case:
1. Install RHEL 9 machine with usbguard package. In my case it was version 1.0.0-16.
2. pick some rule with rhel9 prodtype, e.g. service_usbguard_enabled. set its platform to package[usbguard]>1.0.0
3. ./build_product rhel9
4. examine the datastream. By following various references, you will get to the OVAL state <linux:rpminfo_state id="oval:ssg-ste_inventory_test_package_usbguard_gt_1_0_0_gt_1_0_0_installed:ste:1" version="1">. This state has the following contents: <linux:evr datatype="evr_string" operation="greater than">0:1.0.0-0</linux:evr>.
Let's ignore the fact that the state name is strange for now.
5. remember this part of the file. Now let's go experimenting.
6. evaluate the rule service_usbguard_enabled, or any other rule you chose in the step 1. I expect it to be not applicable, but it results in fail. I believe that the reason is that the release part contains 0, which is lower than 16. If I manually edit the datastream and set it to 17, it starts returning "not applicable".
So we need to solve this somehow. I tried to remove the release part manually - did not help. I tried to changing it to * or z, this did not help. Is there some release character which would be just higher than others?
Proposals:
Document this behavior (although I don't like the  behavior). e.g. if you want to make a rule applicable to package greater than 1.2.3, you have to use > 1.2.4. Or maybe >= 1.2.4, I did not test this case.
OR
Find some default "release" which will be put there instead of "0".
OR
Put release specifications back and document caveats

jan-cerny commented:
@vojtapolasek thanks for opening this topic, I will take a look into it

jan-cerny commented:
I think that the main problem is that when we specify the version condition without specifying the release part and using the upstream versioning, eg. `package[foo]>1.0` then the RPM packaged version of foo 1.0 eg. on RHEL 9 will always satisfy the condition because it will be packaged as `foo-1.0.0-1.el9` and `1.0.0-1.el9` is considered greater than `1.0` by the `rpmvercmp` algorithm. Therefore, people would have to find the first version in which the behavior of the foo component changed, eg. 1.1,  and use that with the greater than or greater than or equal operator eg. `package[foo]>=1.1`. That can be difficult to realize a do it correctly.

Another problem is that when the feature or behavior change  is introduced in a specific release of the package, eg. `foo-1.0.0-7.el9` introduces a new feature, we will not be able to write any platform expressions for that, because `package[foo]>1.0` would match also `foo-1.0.0-1.el9` which doesn't have the feature and `package[foo]>=1.1` wouldn't match `foo-1.0.0-7.el9` which has the feature. We might want to explore the option of having the release part in the expressions to accommodate this use case.

jan-cerny commented:
I have add code to support epochs.

Then, I have add documentation to explain the version caveat.

jan-cerny commented:
/retest

vojtapolasek commented on docs/manual/developer/06_contributing_with_content.md:
```diff
@@ -1240,6 +1240,43 @@ The `platform` property of a rule (or a group) can contain a [Boolean expression
+But, in general, we recommend to derive rule applicability from operating system release version instead of basing that on a version of a specific package.
+The reason for this recommendation is that packages in Linux distributions might be patched, bug fixes and features sometimes get backported to older stable versions, so the behavior of the software might differ from upstream release behavior.
+If the check would be based on operating system version, you can track what packages are distributed there and what patches they contain.
+Examine the actual contents and changelog of the rpm/deb pakcage in the specific product.
```
```suggestion
Examine the actual contents and changelog of the rpm/deb package in the specific product.
```

vojtapolasek replied:
```suggestion
Examine the actual contents and changelog of the rpm/deb package in the specific product.
```

vojtapolasek commented:
@matejak  I did some tests and this looks very promising.
I did the following:
1. install RHEL 9 machine with usbguard package
2. modify rule service_usbguard_enabled
    a) package[usbguard]>=1.0.1 - not applicable
    b)  package[usbguard]<1.0.0 - applicable
    c) package[usbguard]>=1:1.0.0 - not applicable

I tested OVAL, Bash, Ansible. Everything works as described in the documentation.

jan-cerny commented:
@vojtapolasek thank you very much for this testing! it's very helpful :+1: :+1:

jan-cerny commented:
/packit retest

jan-cerny commented:
/packit test

jan-cerny commented:
I have resolved the conflict and rebased the PR on the top of the latest upstream master branch.

jan-cerny commented:
@matejak do you have any opinion?

jan-cerny commented:
I have discovered that the build system internally prepends "0:" to the version string if the content author doesn't specify epoch within his platform expression in his rule.yml. Then, the Bash and Ansible Jinja macro prepend "0:"  to the version string returned by RPM if the package doesn't have epoch. That means that the actual version comparison in the conditionals will be normalized.

I have changed macro bash_compare_version, added a BATS unit test and clarified the documentation regarding Epoch.

jan-cerny commented:
I have rebased the PR on the top of the latest upstream master branch.

jan-cerny commented:
I have improved the version string sanity check in the package platform template.

jan-cerny commented:
I have resolved the conflict and rebased this PR on the top of the latest upstream master branch.

matejak commented:
I generally favor merge of the PR, and I have added a commit that

- changes the ordering of macro arguments, so they are more intuitive, and
- adds test cases that pinpoint certain undefined behavior of the comparison algorithm that should be prevented on higher levels of the code abstraction.

jan-cerny commented:
/retest

jan-cerny commented:
/retest

vojtapolasek commented:
/retest

jan-cerny commented:
/retest

jan-cerny commented:
I have rebased this PR on the top of the latest upstream master branch.

jan-cerny commented:
@matejak ?

jan-cerny commented:
I will repeat again that this PR blocks PRs #10167 and #10209.

jan-cerny commented:
@matejak bump

matejak reviewed (approved):
I like the PR very much, especially after contributing to it.

matejak commented:
Thank you, this contribution will be useful for fine-grained runtime checks.
